src/Components/Endpoints/src/DependencyInjection/HttpNavigationManager.cs

```diff
@@ -7,11 +7,24 @@ namespace Microsoft.AspNetCore.Components.Endpoints;
 
 internal sealed class HttpNavigationManager : NavigationManager, IHostEnvironmentNavigationManager
 {
+    private IEndpointHtmlRenderer? _renderer;
+
     void IHostEnvironmentNavigationManager.Initialize(string baseUri, string uri) => Initialize(baseUri, uri);
 
+    void IHostEnvironmentNavigationManager.Initialize(string baseUri, string uri, IEndpointHtmlRenderer? renderer)
+    {
+        Initialize(baseUri, uri);
+        _renderer = renderer;
+    }
+
     protected override void NavigateToCore(string uri, NavigationOptions options)
     {
         var absoluteUriString = ToAbsoluteUri(uri).AbsoluteUri;
         throw new NavigationException(absoluteUriString);
     }
+
+    protected override void NotFoundCore()
+    {
+        _renderer?.SetNotFoundResponse();
+    }
```

Comment:
Could we avoid the additional `IEndpointHtmlRenderer` and subscribe to the `NavigationManager.NotFound` event from the `EndpointHtmlRenderer` to set a 404 status code when a `NotFound` event is triggered?

(Essentially, reverse the relationship, instead of NavigationManager -> Renderer, Renderer -> NavigationManager) [Here](https://github.com/dotnet/aspnetcore/blob/12d57ddde6bd2f47757364252596cfa89df2ef60/src/Components/Endpoints/src/Rendering/EndpointHtmlRenderer.cs#L79) subscribe to the event to trigger whatever is needed when a 404 happens